### PR TITLE
Restrict size of automatically broadcasted tables

### PR DESCRIPTION
**Pull request by sopel39**
Add join-max-broadcast-table-size property

Property defines maximum size of a table
that can be broadcast for JOIN (0 disables the limit).

Supersedes https://github.com/prestodb/presto/pull/11570

### CODE REVIEW

**Comment by sopel39** [comment]
As it turns out, using upper estimate calculator in current form introduces more problems than it benefits it brings. Here are results comparing:
*

**Comment by sopel39** [comment]
Observations:
1) upper estimate calculator does bring more problems than benefits
2) upper estimate calculator improves performance (vs disabled and current "probabilistic" stats calculator) for queries: tpcds/q22, tpcds/q72, tpcds/q95, tpcds/q76. This suggest that for those queries we are overestimating number of rows for build side.
3) using current "probabilistic" calculator with broadcast table size limit improves performance (vs no limit) in queries: tpch/q09, tpcds/q80, tpcds/q81, tpcds/q64, tpcds/q40, tpcds/q24_1, tpcds/q24_2
4) using current "probabilistic" calculator with broadcast table size limit degrades performance for some queries: tpch/q09, tpcds/q82, tpcds/q72, tpcds/q85, tpcds/q43, tpcds/q50, tpcds/q37.
Overall (weighted by query time) improvement is 0.11x
5) We greatly underestimate join build side construction, see query tpch/09. Will create separate issue for that

Conclusions:
* I would land toggle for limiting broadcast build side but using "normal" stats calculator
* Limiting broadcast table size improves cluster memory utilization and concurrency. It can also greatly improve performance for queries and prevent them failing with hitting memory limit. Such flag is used by other engines (e.g: spark, Netizza)

Next steps:
* Fix cost calculator to better account for join build side construction (related to: https://github.com/prestodb/presto/issues/11615)
* Improve (if possible) stats calculator so that it doesn't underaccount number of rows (see 2) from "observations")
* Maybe improve upper stats calculator as it helped with some queries

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 1]
```diff
@@ -149,6 +151,15 @@ public SystemSessionProperties(
                         false,
                         value -> JoinDistributionType.valueOf(((String) value).toUpperCase()),
                         JoinDistributionType::name),
+                new PropertyMetadata<>(
+                        JOIN_MAX_BROADCAST_TABLE_SIZE,
+                        "Maximum size of a table that can be broadcast for JOIN (0 disables the limit). The most pessimistic size estimate is taken when doing comparison.",
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: `The most pessimistic size estimate is taken when doing comparison.` - This doesn't seem to be the case, is it?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java [hunk 2]
```diff
@@ -466,4 +465,83 @@ public void testFlipAndReplicateRightOuterJoin()
...
+        assertDetermineJoinDistributionType()
+                .setSystemProperty(JOIN_DISTRIBUTION_TYPE, JoinDistributionType.AUTOMATIC.name())
+                .setSystemProperty(JOIN_MAX_BROADCAST_TABLE_SIZE, "100MB")
+                .overrideStats("valuesA", probeSideStatsEstimate)
+                .overrideStats("valuesB", buildSideStatsEstimate)
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.values(new PlanNodeId("valuesA"), aRows, p.symbol("A1", BIGINT)),
+                                p.filter(TRUE_LITERAL, p.values(new PlanNodeId("valuesB"), bRows, p.symbol("B1", BIGINT))),
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: Why not just `p.values`?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java [hunk 3]
```diff
@@ -466,4 +465,83 @@ public void testFlipAndReplicateRightOuterJoin()
...
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: Ditto about filter

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestReorderJoins.java [hunk 4]
```diff
@@ -418,6 +422,75 @@ public void testSmallerJoinFirst()
...
+
+        // B table is small enough to be replicated in AUTOMATIC_RESTRICTED mode
+        assertReorderJoins()
+                .setSystemProperty(JOIN_DISTRIBUTION_TYPE, AUTOMATIC.name())
+                .setSystemProperty(JOIN_MAX_BROADCAST_TABLE_SIZE, "100MB")
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.values(new PlanNodeId("valuesA"), aRows, p.symbol("A1")),
+                                p.filter(TRUE_LITERAL, p.values(new PlanNodeId("valuesB"), bRows, p.symbol("B1"))),
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestReorderJoins.java]: ditto

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestReorderJoins.java [hunk 5]
```diff
@@ -418,6 +422,75 @@ public void testSmallerJoinFirst()
...
+
+        // B table exceeds AUTOMATIC_RESTRICTED limit therefore it is partitioned
+        assertReorderJoins()
+                .setSystemProperty(JOIN_DISTRIBUTION_TYPE, AUTOMATIC.name())
+                .setSystemProperty(JOIN_MAX_BROADCAST_TABLE_SIZE, "100MB")
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.values(new PlanNodeId("valuesA"), aRows, p.symbol("A1")),
+                                p.filter(TRUE_LITERAL, p.values(new PlanNodeId("valuesB"), bRows, p.symbol("B1"))),
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestReorderJoins.java]: Ditto

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: > 0 disables the limit

It should be rather `+INF` to disable limit

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 6]
```diff
@@ -28,6 +28,7 @@
 import javax.inject.Inject;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Please extend the commit message. Explain that it is about CBO and explain the problem and when it should be used.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: > The most pessimistic size estimate is taken when doing comparison.

I would prefer to not to go to such details.

> Maximum size

Please update to `Maximum estimated size`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 7]
```diff
@@ -68,20 +72,31 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
...
         CostProvider costProvider = context.getCostProvider();
         List<PlanNodeWithCost> possibleJoinNodes = new ArrayList<>();
 
-        if (!mustPartition(joinNode)) {
+        if (!mustPartition(joinNode) && canReplicate(joinNode, context)) {
             possibleJoinNodes.add(getJoinNodeWithCost(costProvider, joinNode.withDistributionType(REPLICATED)));
         }
         if (!mustReplicate(joinNode, context)) {
             possibleJoinNodes.add(getJoinNodeWithCost(costProvider, joinNode.withDistributionType(PARTITIONED)));
         }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: above two `ifs` can be extracted as method and reused below

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: It's leftover from upper boundary stats calculator.

**Comment by sopel39** [comment]
@mbasmanova that's the one (`autoBroadcastJoinThreshold`)

**Pull request merged**